File: 2023/07.13/송기훈/개미굴.java
package baekjoon.gold.three;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.TreeMap;

public class 개미굴 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int N;


    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());

        TreeMap<String, Object> root = new TreeMap<>();

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int K = Integer.parseInt(st.nextToken());

            TreeMap<String, Object> targetNode = root;

            for (int j = 0; j < K; j++) {
                String node = st.nextToken();

                if (!targetNode.containsKey(node)) {
                    targetNode.put(node, new TreeMap<>());
                }

                targetNode = (TreeMap<String, Object>) targetNode.get(node);

            }
        }
        printResult(root, 0);

    }

    static void printResult(TreeMap<String, Object> treeMap, int depth) {
        for (Object object : treeMap.keySet()) {
            for (int i = 0; i < depth; i++) {
                System.out.print("--");
            }
            System.out.println(object);
            printResult((TreeMap<String, Object>) treeMap.get(object), depth + 1);
        }
    }

}

File: 2023/07.13/조수연/개미굴.kt
import java.io.*
import java.util.*
import kotlin.collections.HashMap

data class Node(val children: HashMap<String, Node> = HashMap())

val bw = BufferedWriter(OutputStreamWriter(System.out))

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val n = readLine().toInt()
    val root = Node()

    repeat(n) {
        val token = StringTokenizer(readLine())
        val k = token.nextToken().toInt()
        var current = root

        repeat(k) {
            val word = token.nextToken()
            if (!current.children.containsKey(word)) { // 해당 층에 word가 없으면 추가
                current.children[word] = Node()
            }
            current = current.children[word]!! // 층 갱신
        }
    }

    printNode(root, "")

    bw.flush()
    bw.close()
}

fun printNode(root: Node, bars: String) {
    val keys = root.children.keys.sorted()

    for (i in keys.indices) {
        bw.write("$bars${keys[i]}\n")
        printNode(root.children[keys[i]]!!, "$bars--")
    }
}
